Source/ASNetworkImageNode.mm

```diff
@@ -352,32 +369,33 @@ - (void)didExitPreloadState
     if (_imageWasSetExternally) {
       return;
     }
-
-    [self _cancelDownloadAndClearImage];
   }
+
+  [self _cancelDownloadAndClearImage];
 }
 
 - (void)didEnterPreloadState
 {
   [super didEnterPreloadState];
   
-  {
-    ASDN::MutexLocker l(__instanceLock__);
-    // Image was set externally no need to load an image
-    [self _lazilyLoadImageIfNecessary];
-  }
+  // Image was set externally no need to load an image
+  [self _lazilyLoadImageIfNecessary];
 }
 
 #pragma mark - Private methods, safe to call without lock
 
 - (void)handleProgressImage:(UIImage *)progressImage progress:(CGFloat)progress downloadIdentifier:(nullable id)downloadIdentifier
 {
-  ASDN::MutexLocker l(__instanceLock__);
-  // Getting a result back for a different download identifier, download must not have been successfully canceled
-  if (ASObjectIsEqual(_downloadIdentifier, downloadIdentifier) == NO && downloadIdentifier != nil) {
-    return;
+  {
+    ASDN::MutexLocker l(__instanceLock__);
```

Comment:
This change seems like it likely won't cause a problem, but it seems theoretically possible. If the setImage for a progress image gets the after the setImage for the download completion we could theoretically end up with the progress image and never get the complete image.